public/app/features/explore/hooks/useStateSync.ts

```diff
@@ -0,0 +1,409 @@
+import { isEmpty, isEqual, isObject, mapValues, omitBy } from 'lodash';
+import { useEffect, useRef } from 'react';
+
+import {
+  CoreApp,
+  serializeStateToUrlParam,
+  ExploreUrlState,
+  isDateTime,
+  TimeRange,
+  RawTimeRange,
+  DataSourceApi,
+} from '@grafana/data';
+import { DataQuery, DataSourceRef } from '@grafana/schema';
+import { useGrafana } from 'app/core/context/GrafanaContext';
+import { clearQueryKeys, getLastUsedDatasourceUID, parseUrlState } from 'app/core/utils/explore';
+import { getDatasourceSrv } from 'app/features/plugins/datasource_srv';
+import { MIXED_DATASOURCE_NAME } from 'app/plugins/datasource/mixed/MixedDataSource';
+import { addListener, ExploreId, ExploreItemState, ExploreQueryParams, useDispatch, useSelector } from 'app/types';
+
+import { changeDatasource } from '../state/datasource';
+import { initializeExplore } from '../state/explorePane';
+import { clearPanes, splitClose, splitOpen, syncTimesAction } from '../state/main';
+import { runQueries, setQueriesAction } from '../state/query';
+import { selectPanes } from '../state/selectors';
+import { changeRangeAction, updateTime } from '../state/time';
+import { DEFAULT_RANGE } from '../state/utils';
+import { withUniqueRefIds } from '../utils/queries';
+
+/**
+ * Bi-directionally syncs URL changes with Explore's state.
+ */
+export function useStateSync(params: ExploreQueryParams) {
+  const {
+    location,
+    config: {
+      featureToggles: { exploreMixedDatasource },
+    },
+  } = useGrafana();
+  const dispatch = useDispatch();
+  const statePanes = useSelector(selectPanes);
+  const orgId = useSelector((state) => state.user.orgId);
+  const prevParams = useRef<ExploreQueryParams>(params);
+  const initState = useRef<'notstarted' | 'pending' | 'done'>('notstarted');
+
+  useEffect(() => {
+    // This happens when the user navigates to an explore "empty page" while within Explore.
+    // ie. by clicking on the explore when explore is active.
+    if (!params.left && !params.right) {
```

Comment:
Should we also check for keys in `panes` param?

Reply:
the `panes` params doesn't exist yet, it will be introduced in a follow up PR alongside the migrations.

Reply:
Coolio, just asking because there was already some code for handling pane keys.

Reply:
the other `panes` key used in the PR is referencing the panes object in redux, this one is referencing the query params